### ISSUE TITLE
Ideas for geojson / GeoJsonDataSource / multipolygons integrations

### ISSUE THREAD

**Issue by jsignell** [bokeh/bokeh#8359]
Since `MultiPolygons` has landed and is closely modeled on geojson, we can do a better job with `GeoJsonDataSource` in the case of MultiPolygons.

## Currently

`GeoJsonDataSource` converts  any `MultiPolygon` to a `NaN` separated list of x and y coordinates for the exterior ring. This maps nicely to `Patches` but any holes are ignored. 

## Proposal

`GeoJsonDataSource` should leave `MultiPolygons` in their nested structure for use with the `MultiPolygons` glyph. I guess `zs` can be one degree less nested because holes have no depth.

A new test could look like:

```coffee
  describe "xy MultiPolygon", ->
    # This has two polygons, and the second one has has a hole in it
    geojson = """{
      "type": "GeometryCollection",
      "geometries": [{
        "type": "MultiPolygon",
        "coordinates": [
          [[ [102.0, 2.0], [103.0, 2.1] ]],
          [[ [100.0, 0.0], [101.0, 0.1] ],
           [ [100.2, 0.2], [100.8, 0.3] ]]
        ]
        }
      ]
    }"""
    geo = new GeoJSONDataSource({geojson: geojson})

    it "should add 3 times nested lists of xs, ys and 2 times nested lists of zs", ->
      expected_data = {'x': [NaN], 'y': [NaN], 'z': [NaN], 'xs':[[[[102.0, 103.0]], [[100.0, 101.0], [100.2, 100.8]]]], 'ys':[[[[2.0, 2.1]], [[0, 0.1], [0.2, 0.3]]]], 'zs': [[[NaN, NaN], [NaN, NaN]]]}
      expect(geo.data).to.be.deep.equal(expected_data)

```

**Comment by bryevdv** [comment]
Although it kind of sucks, we probably want to add a flag for some mode "patches" vs "multipolygons". We could leave the default as "patches" or possibly be clever and let the glyph set things (if the conversion can be deferred)

**Comment by gmerritt123** [comment]
Gentle bump on this, and sharing my learning/implementation... --> I've managed to build a workaround implementation for a major feature on my geologic XS app but holy moly it was a PITA. Arbitrary nodal info (i.e. finite element process variables, in my case hydraulic pressure or contaminant concentrations), gets interpolated to a user-defined cross section, that irregularly shaped data gets d3 delaunay triangulated in order to grid it at an appropriate resolution (i.e. a JS implementation of scipy's griddata), then that gridded data gets d3 contoured (marching squares contouring). d3's contour routine puts results out in geojson multipolygons, which a) can have holes, but also b) are overlapping (e.g. the 250 contour lives exclusively inside the 120 contour, etc.). I had to resort to using https://github.com/bjornharrtell/jsts to a) clip the contours to the xs geometry and b) clip the 250 out of the 120 etc. So I had that all working on the JS side nicely, all that was left was to transform the geojson (or jsts object) back into bokeh-cds ready multipolygons, but found that I could not instantiate a GeoJSONDataSource on the JS side using geojson containing holes --> Little fiddle demo here: https://jsfiddle.net/7vjan4k3/

![Image](https://github.com/user-attachments/assets/47e96f6a-741e-4f99-914a-a8f85832e0ea)

I ended up formulating two functions to do this for me (one to build patch/multiline-like CDS, and another for a MultiPolygon-like CDS), which bypasses using GeoJSONDataSource entirely:

```
//takes a geojson spec multipolygon and converts to bokeh multi-polygon glyph xs and ys
function convertGeoJSONToBokehMultiPolygons(geometry) {
  const xs = [];
  const ys = [];

  if (geometry.type === "MultiPolygon") {
    const multipolygon_xs = [];
    const multipolygon_ys = [];

    geometry.coordinates.forEach(polygon => {
      const polygon_xs = [];
      const polygon_ys = [];

      polygon.forEach(ring => {
        const x = ring.map(coord => coord[0]);
        const y = ring.map(coord => coord[1]);
        polygon_xs.push(x);
        polygon_ys.push(y);
      });

      multipolygon_xs.push(polygon_xs);
      multipolygon_ys.push(polygon_ys);
    });

    xs.push(multipolygon_xs);
    ys.push(multipolygon_ys);
  }

  else if (geometry.type === "Polygon") {
    const polygon_xs = [];
    const polygon_ys = [];

    geometry.coordinates.forEach(ring => {
      const x = ring.map(coord => coord[0]);
      const y = ring.map(coord => coord[1]);
      polygon_xs.push(x);
      polygon_ys.push(y);
    });

    xs.push([polygon_xs]); // wrap in an extra array to match MultiPolygon structure
    ys.push([polygon_ys]);
  }

  else {
    throw new Error("Geometry must be of type Polygon or MultiPolygon");
  }

  return { xs, ys };
}

//takes a geojson spec multipolygon and converts to bokeh multiline or patches glyph xs and ys
function convertGeoJSONToBokehPatches(geometry) {
  const xs = [];
  const ys = [];
  if (geometry.type === "MultiPolygon") {
    geometry.coordinates.forEach(polygon => {
      polygon.forEach(ring => {
        const x = [];
        const y = [];
        ring.forEach(coord => {
          x.push(coord[0]);
          y.push(coord[1]);
        });
        xs.push(x);
        ys.push(y);
      });
    });
  } else if (geometry.type === "Polygon") {
    geometry.coordinates.forEach(ring => {
      const x = [];
      const y = [];
      ring.forEach(coord => {
        x.push(coord[0]);
        y.push(coord[1]);
      });
      xs.push(x);
      ys.push(y);
    });
  } else {
    throw new Error("Geometry must be of type Polygon or MultiPolygon");
  }
  return { xs, ys };
}
```
And it works:

![Image](https://github.com/user-attachments/assets/a8636461-706b-4454-a069-95117765642a)

And on the "build side" I'm definitely doing what @bryevdv suggests with a "flag":

![Image](https://github.com/user-attachments/assets/161318c5-ac21-44ba-942a-15f810518686)

So when building a site/project specific application I can decide if I want the contours to be lines (typically to show the water table) or clipped-out multipolygons (contaminant concentrations).

Anyway, that's my 2c on what I've implemented to get this working for me :-)

**Comment by bryevdv** [comment]
I'm pretty far removed from looking at any of this closely, and I don't actually have any concrete code / use-case to bang against for development, so I'm not sure I would trust my instincts, re: requirements. I will say Bokeh 4.0 is at least starting to be thought about so that would be an ideal time for any breaking changes, if breaking changes are needed. 

I will say I could probably be very helpful starting from a working POC to get it into whatever final shape is needed to merge (docs, tests, any special bokehjs coding considerstions, etc). @gmerritt123 do you think you could bang something concrete together that makes like easier in the way you are looking for it to be easier, that we can then iterate on?

**Comment by gmerritt123** [comment]
I think the friction lies with the fact that GeoJSON feature collections can contain multiple geometry types and the core Bokeh paradigm is driving one glyph instance with one CDS instance. Bokeh currently expects to receive a feature collection when instantiating a GeoJSONDataSource, but a feature collection can contain multiple geometry types, e.g. Points, Lines, Polygons, and their "multi" versions as well. Obviously that won't work very well if we want to be able to convert GeoJSON to some "CDS" ready format. Making even more friction, GeoJSON uses coordinate pairs but bokeh CDS separates out x and y into separate columns (so a lot of transposition has to happen to do the transform), and this becomes a convoluted mess with a multipolygon with holes.

This is also discussed here --> https://github.com/bokeh/bokeh/issues/3708 and also here -->https://github.com/bokeh/bokeh/wiki/Polygons-Patches-with-Holes

Prefacing this with the disclaimer that I'm not a developer, but my immediate "path of least resistance/breaking changes" suggestions would be as follows:

1) Immediately update this part of the docs:

![Image](https://github.com/user-attachments/assets/125f0e78-075f-47f1-bf77-e3131a353a2c)

To say something along the lines of "the features passed must all be one geometry type. Currently, Point, Lines, and MultiLines are supported. Polygona and MultiPolygon support is limited: exterior rings will be converted to xs and ys fields in a ColumnDataSource, but holes will not be represented."

2) Future feature request would be to "fully" handle Polygon and MultiPolygon geojson, perhaps using my function above, to specifically convert MultiPolygon and Polygon geojson to a bokeh-glyph `MultiPolygon` ready xs and ys columns. I think everything else could stay the same?

3) Update the docs to state these "conversion rules" specifically: Points will be converted to x and y columns (i.e. ideal for Scatter glyph), Lines and MultiLine to xs and ys columns (i.e. ideal for multiline or patches glyph), and Polygons/MultiPolygons will be converted to xs and ys columns suitable for the MultiPolygon glyph.

The more ambitious/breaking change would be to start delving into glyphs/renderers accepting different data structures, like not only geojson but also things like "index_source" as discussed here --> https://github.com/bokeh/bokeh/issues/14139#issuecomment-2456737456

But I'm not sure where the best place to start on that would be. Hopefully my 2c here helps.

**Comment by bryevdv** [comment]
So a little historical context may be enlightening here. At the time `GeoJSONDataSource` was added, `MultiPolygons` did not exists and most/all glyphs just expected simple lists of points. So, that was the emphasis on having `GeoJSONDataSource` adapt things to the "standard" CDS format. I kind of think 4.0 could be an opportunity to re-imagine things.

*Extremely* off-the-cuff idea since I have not given any proper thought to this yet. Perhaps `GeoJSONDataSource` could expose dedicated fields `point`, `line`, `polygon`, `multi_point`, `multi_line`, `multi_polygon`. Then a few glyphs can be taught how to use the appropriate field for them when the data source is a geo source. 

Question to think about would be e.g. what to do with collections (punt?) and where how to store extra data fields... 

Alternatively, maybe a `GeoJSON` model literally creates sub-CDS's for each geometry, so you'd be more explicit:
```python
geo = GeoJSON(...)

p.line(..., source=geo.line)
p.scatter(..., source=geo.point)
```

That second approach would maybe require very little change on the BokehJS side but it would make things less reactive in a sense (i.e. I am not sure how well this would work to support updating GeoJSON after it's been created, either from Python or from JS... but then, is that usually not a thing anyway?)

Edit: I guess that is a really good clarifying question. Would it be acceptable for GeoJSON *geometry* data to be more or less read-only after ingestion? i.e. changing associated attributes OK, but the geometry itself can't really be updated "in place"? Or if we do need to support updating geometries then some concrete motivating examples would help.

**Comment by gmerritt123** [comment]
Yeah, it was definitely obvious to me that MultiPolygon glyph came after GeoJSONDataSource when I dug deeply into this, but the docs currently don't make it clear that multipolygon/polygon GeoJSON is not properly supported.

I don't mind your latter idea of a GeoJSON model "ingesting" geojson data and storing that data on CDS instances for each geometry type.

> " it would make things less reactive in a sense (i.e. I am not sure how well this would work to support updating GeoJSON after it's been created, either from Python or from JS... but then, is that usually not a thing anyway?)"

Maybe I don't follow what you're saying here, but I think I can stand up and say that this definitely does matter/should be available. For my use case above, I'm using external JS libraries to generate geojson based on some underlying dataset and user action, then i'm converting that geojson to "bokeh-ready" CDS to update a CDS and the renderers they drive. I can think of other kinda obvious applications - there are a tonne of REST API based calls to things like Arc Maps (GIS stuff) that return geojson on request: any bokeh app wanting to do a JS side request to get new geojson from a server would benefit from JS-side "auto-ingestion". Way I see it working:

```
#instantiate a GeoJSON model python side while building the app, with no args at all:
gj = GeoJSON()
#this instantiation automatically populates a bunch of CDS instances inside it even if you don't supply initial geojson data
gj.line--> ColumnDataSource(id='p1') #ready for multiline
gj.point--> ColumnDataSource(id='p2') #ready for scatter
gj.polygon --> ColumnDataSource(id='p3') #ready for multipolygon glyph
```

Then over on JS side have something like:
```
gj_data = go_get_new_geojson_from_somewhere()

// pass that new gj data to the GeoJSON model
gj.geojson = gj_data
//then either automatically, or via some method, tell the GeoJSON model to update its CDS's:
gj.update_sources() //this would facilitate js-side "ingestion" and the source's data would get updated, which would propagate to the renderers
```

Hopefully that makes sense?
